File: clientapp/lib/domain/api_client.dart
import 'package:dio/dio.dart';
import 'package:servicegeniestore/domain/ad/ad.dart';
import 'package:servicegeniestore/domain/auth/auth_response.dart';
import 'package:servicegeniestore/domain/order/order.dart';
import 'package:servicegeniestore/domain/order/order_create_dto.dart';
import 'package:servicegeniestore/domain/service/date_time_slots.dart';
import 'package:servicegeniestore/domain/service/service.dart';
import 'package:servicegeniestore/domain/shared/category.dart';
import 'package:servicegeniestore/domain/shared/coinbase_charge.dart';
import 'package:servicegeniestore/domain/shared/currency.dart';
import 'package:servicegeniestore/domain/shared/payment_method.dart';
import 'package:servicegeniestore/domain/user/user.dart';

class ApiClient {
  final Dio _dio;
  final String baseURL;
  String? accessToken = null;

  ApiClient(this.baseURL) : _dio = Dio();

  Future<AuthResponse> login(String email, String password) async {
    try {
      final response = await _dio.post(
        '$baseURL/api/authentication/login',
        data: {
          'email': email,
          'password': password,
        },
      );
      return AuthResponse.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to login: $error');
    }
  }

  Future<User> getUser() async {
    try {
      final response = await _dio.get('$baseURL/api/user/currentUser',
          queryParameters: {"access_token": accessToken});
      return User.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch services: $error');
    }
  }

  Future<List<Service>> getServices() async {
    try {
      final response = await _dio.get('$baseURL/api/service');
      final List<dynamic> data = response.data;
      return data.map((json) => Service.fromJson(json)).toList();
    } catch (error) {
      throw Exception('Failed to fetch services: $error');
    }
  }

  Future<Service> getServiceById(String id) async {
    try {
      final response = await _dio.get('$baseURL/api/service/$id');
      return Service.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch service: $error');
    }
  }

  Future<List<DateTimeSlots>> getServiceAvailableTimeSlots(String id) async {
    try {
      final response =
          await _dio.get('$baseURL/api/service/$id/available-slots');
      final List<dynamic> data = response.data;
      return data.map((json) => DateTimeSlots.fromJson(json)).toList();
    } catch (error) {
      throw Exception('Failed to fetch service available time slots: $error');
    }
  }

  Future<User> getUserById(String id) async {
    try {
      final response = await _dio.get('$baseURL/api/user/$id');
      return User.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch user: $error');
    }
  }

  Future<List<Ad>> getAds() async {
    try {
      final response = await _dio.get('$baseURL/api/ad');
      final List<dynamic> data = response.data;
      return data.map((json) => Ad.fromJson(json)).toList();
    } catch (error) {
      throw Exception('Failed to fetch ads: $error');
    }
  }

  Future<void> updateAd(String id, Map<String, dynamic> updatedAdData) async {
    try {
      await _dio.put('$baseURL/api/ad/$id', data: updatedAdData);
    } catch (error) {
      throw Exception('Failed to update ad: $error');
    }
  }

  Future<List<PaymentMethod>> getPaymentMethods() async {
    try {
      final response = await _dio.get('$baseURL/api/payment-method');
      final List<dynamic> data = response.data;
      return data.map((json) => PaymentMethod.fromJson(json)).toList();
    } catch (error) {
      throw Exception('Failed to fetch payment methods: $error');
    }
  }

  Future<PaymentMethod> getPaymentMethod(String paymentMethodId) async {
    try {
      final response =
          await _dio.get('$baseURL/api/payment-method/$paymentMethodId');
      return PaymentMethod.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch payment method: $error');
    }
  }

  Future<PaymentMethod> getPaymentMethodById(String id) async {
    try {
      final response = await _dio.get('$baseURL/api/payment-method/$id');
      return PaymentMethod.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch payment method: $error');
    }
  }

  Future<List<ServiceCategory>> getCategories() async {
    try {
      final response = await _dio.get('$baseURL/api/category');
      final List<dynamic> data = response.data;
      return data.map((json) => ServiceCategory.fromJson(json)).toList();
    } catch (error) {
      throw Exception('Failed to fetch categories: $error');
    }
  }

  Future<ServiceCategory> getCategoryById(String id) async {
    try {
      final response = await _dio.get('$baseURL/api/category/$id');
      return ServiceCategory.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch category: $error');
    }
  }

  Future<void> updateCategory(
      String id, Map<String, dynamic> updatedCategoryData) async {
    try {
      await _dio.put('$baseURL/api/category/$id', data: updatedCategoryData);
    } catch (error) {
      throw Exception('Failed to update category: $error');
    }
  }

  Future<List<Currency>> getCurrencies() async {
    try {
      final response = await _dio.get('$baseURL/api/currency');
      final List<dynamic> data = response.data;
      return data.map((json) => Currency.fromJson(json)).toList();
    } catch (error) {
      throw Exception('Failed to fetch currency: $error');
    }
  }

  Future<Currency> getCurrencyById(String id) async {
    try {
      final response = await _dio.get('$baseURL/api/currency/$id');
      return Currency.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch currency: $error');
    }
  }

  Future<void> createOrder(OrderCreateDto orderData) async {
    try {
      await _dio.post('$baseURL/api/order', data: orderData);
    } catch (error) {
      throw Exception('Failed to create order: $error');
    }
  }

  Future<Order> getOrderById(String id) async {
    try {
      final response = await _dio.get('$baseURL/api/order/$id');
      return Order.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch order: $error');
    }
  }

  Future<List<Order>> getOrdersByUserId(String userId) async {
    try {
      final response = await _dio.get('$baseURL/api/order/user/$userId');
      final List<dynamic> data = response.data;
      return data.map((json) => Order.fromJson(json)).toList();
    } catch (error) {
      throw Exception('Failed to fetch order: $error');
    }
  }

  Future<void> updateOrder(
      String id, Map<String, dynamic> updateOrderData) async {
    try {
      await _dio.put('$baseURL/api/order/$id', data: updateOrderData);
    } catch (error) {
      throw Exception('Failed to update order: $error');
    }
  }

  Future<Ad> getAdById(String id) async {
    try {
      final response = await _dio.get('$baseURL/api/ad/$id');
      return Ad.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to fetch ad: $error');
    }
  }

  Future<CoinbaseCharge> createCharge(
      String serviceId, String currencyId) async {
    try {
      final response = await _dio.post('$baseURL/api/coinbase/create-charge',
          queryParameters: {"serviceId": serviceId, "currencyId": currencyId});
      return CoinbaseCharge.fromJson(response.data);
    } catch (error) {
      throw Exception('Failed to create charge: $error');
    }
  }
}

File: admin-dashboard-remix/app/services/ApiClient.ts
import { Category } from "@/domain/models/Category";
import { PaymentMethod } from "@/domain/models/PaymentMethod";
import { Service } from "@/domain/services/Service";
import { Order } from "@/domain/models/Order";
import { Currency } from "@/domain/models/Currency";

interface AuthResponse {
  accessToken: string;
}

class APIClient {
  private readonly baseURL: string;
  accessToken: string | null;

  constructor(baseURL: string) {
    this.baseURL = baseURL;
    this.accessToken = null;
  }

  async login(email: string, password: string): Promise<AuthResponse> {
    const response = await fetch(`${this.baseURL}/api/authentication/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ email: email, password: password }),
    });
    const data = await response.json();
    console.log("Response is", data);
    if (data.accessToken) {
      this.accessToken = data.accessToken;
    }
    return data;
  }

  async getServices(): Promise<Service[]> {
    return this._authenticatedRequest("/api/service");
  }

  async createService(serviceData: any): Promise<any> {
    return this._authenticatedRequest("/api/service", {
      method: 'POST',
      body: JSON.stringify(serviceData)
    });
  }

  async getServiceById(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/service/${id}`);
  }

  async updateService(id: string, updatedServiceData: any): Promise<any> {
    return this._authenticatedRequest(`/api/service/${id}`, {
      method: 'PUT',
      body: JSON.stringify(updatedServiceData)
    });
  }

  async deleteService(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/service/${id}`, {
      method: 'DELETE'
    });
  }

  async getUserById(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/user/${id}`);
  }

  async getUsers(): Promise<any> {
    return this._authenticatedRequest(`/api/user/}`);
  }

  async createUser(userData: any): Promise<any> {
    return this._authenticatedRequest(`/api/user`, {
      method: 'POST',
      body: JSON.stringify(userData)
    });
  }

  async deleteUser(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/user`, {
      method: "DELETE",
    });
  }

  async getAds(): Promise<any> {
    return this._authenticatedRequest(`/api/ad`);
  }

  async createAd(adData: any): Promise<any> {
    return this._authenticatedRequest(`/api/ad`, {
      method: 'POST',
      body: JSON.stringify(adData)
    });
  }

  async updateAd(id: string, updatedAdData: any): Promise<any> {
    return this._authenticatedRequest(`/api/ad/${id}`, {
      method: 'PUT',
      body: JSON.stringify(updatedAdData)
    });
  }
  async deleteAd(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/ad/${id}`, {
      method: "DELETE",
    });
  }

  async getPaymentMethods(): Promise<any> {
    return this._authenticatedRequest("/api/payment-method");
  }

  async getPaymentMethod(paymentMethodId: string): Promise<PaymentMethod> {
    return this._authenticatedRequest(`/api/payment-method/${paymentMethodId}`);
  }

  async createPaymentMethod(paymentMethodData: any): Promise<any> {
    return this._authenticatedRequest("/api/payment-method", {
      method: "POST",
      body: JSON.stringify(paymentMethodData),
    });
  }

  async getPaymentMethodById(id: string): Promise<PaymentMethod> {
    return this._authenticatedRequest(`/api/payment-method/${id}`);
  }

  async updatePaymentMethod(id: string, updatedPaymentMethodData: any): Promise<any> {
    return this._authenticatedRequest(`/api/payment-method/${id}`, {
      method: "PUT",
      body: JSON.stringify(updatedPaymentMethodData),
    });
  }

  async deletePaymentMethod(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/payment-method/${id}`, {
      method: "DELETE",
    });
  }

  async getCategories(): Promise<any> {
    return this._authenticatedRequest("/api/category");
  }

  async createCategory(categoryData: any): Promise<any> {
    return this._authenticatedRequest("/api/category", {
      method: "POST",
      body: JSON.stringify(categoryData),
    });
  }

  async getCategoryById(id: string): Promise<Category> {
    return this._authenticatedRequest(`/api/category/${id}`);
  }

  async updateCategory(id: string, updatedPaymentMethodData: any): Promise<any> {
    return this._authenticatedRequest(`/api/category/${id}`, {
      method: "PUT",
      body: JSON.stringify(updatedPaymentMethodData),
    });
  }

  async deleteCategory(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/category/${id}`, {
      method: "DELETE",
    });
  }

  async getCurrencies(): Promise<any> {
    return this._authenticatedRequest("/api/currency");
  }

  async createCurrency(categoryData: any): Promise<any> {
    return this._authenticatedRequest("/api/currency", {
      method: "POST",
      body: JSON.stringify(categoryData),
    });
  }

  async getCurrencyById(id: string): Promise<Currency> {
    return this._authenticatedRequest(`/api/currency/${id}`);
  }

  async updateCurrency(id: string, updateCurrencyData: any): Promise<any> {
    return this._authenticatedRequest(`/api/currency/${id}`, {
      method: "PUT",
      body: JSON.stringify(updateCurrencyData),
    });
  }

  async getOrders(): Promise<Order[]> {
    return this._authenticatedRequest("/api/order");
  }

  async createOrder(orderData: any): Promise<any> {
    return this._authenticatedRequest("/api/order", {
      method: "POST",
      body: JSON.stringify(orderData),
    });
  }

  async getOrderById(id: string): Promise<Order> {
    return this._authenticatedRequest(`/api/order/${id}`);
  }

  async updateOrder(id: string, updateOrderData: any): Promise<any> {
    return this._authenticatedRequest(`/api/order/${id}`, {
      method: "PUT",
      body: JSON.stringify(updateOrderData),
    });
  }

  async deleteOrder(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/order/${id}`, {
      method: "DELETE",
    });
  }

  async getAdById(id: string): Promise<any> {
    return this._authenticatedRequest(`/api/ad/${id}`);
  }


  async uploadImage(formData: FormData): Promise<any> {
    return await fetch("/uploads", {
      method: "POST",
      body: formData,
    });
  }
  private async _authenticatedRequest(url: string, options: RequestInit = {}): Promise<any> {
    if (!this.accessToken) {
      throw new Error('Access token is missing');
    }
    const headers = options.headers || {};
    const authenticatedOptions = {
      ...options,
      headers: {
        ...headers,
        "Authorization": `Bearer ${this.accessToken}`,
        "Content-Type": "application/json",
      }
    };
    const response = await fetch(`${this.baseURL}${url}`, {
      ...authenticatedOptions,
    });
    if (response.status === 204) {
      return response.status;
    } else {
      return response.json();
    }
  }
}


const apiClient = new APIClient("http://localhost:5000")

export default apiClient